README.md

```diff
@@ -45,3 +45,21 @@ The following steps should serve as a guide when adding new packages or dependen
 - Add a patch if necessary to `patches` directory.
 - If it's a system dependency, add a build rule to `install.sh`. According to the library, it may be necessary to use `build_library` script or adding a special rule to `build_library_with_toolchain`.
 - That's it! Test your build and watch out build errors. Using `verbose` flags when building should help finding hints about the problems that may arise.
+
+## Building your own apps and libraries on top of ROS cross-compiled workspace
+
+You can use `build_catkin_workspace` as a standalone script to build your Android application or library as a catkin package,
+making use of the ROS cross compiled libraries and system dependencies. The examples under [example_workspace](https://github.com/Intermodalics/ros_android/tree/kinetic/example_workspace/src) can be used as a starting point and as a reference to build your code; in short they are Gradle projects wrapped as Catkin packages.
+
+Here's some guidelines to take into account when creating new projects, using `hello_world` app as an example:
+- Your code should have the structure of a ROS Android package, with a top level `CMakeLists` to make Gradle interact with Catkin. Take [this CMakeLists file](https://github.com/Intermodalics/ros_android/blob/kinetic/example_workspace/src/hello_world_example_app/CMakeLists.txt) as an example.
+- A [package.xml](https://github.com/Intermodalics/ros_android/blob/kinetic/example_workspace/src/hello_world_example_app/package.xml) has to be created, declaring Catkin as the build tool, and the ROS dependencies.
+- In the [bottom level gradle buildscript](https://github.com/Intermodalics/ros_android/blob/kinetic/example_workspace/src/hello_world_example_app/app/build.gradle), add the necessary arguments to the `externalNativeBuild` block, and specify ABI filters as in [here](https://github.com/Intermodalics/ros_android/blob/kinetic/example_workspace/src/hello_world_example_app/app/build.gradle#L10-L23). Then, specify the path to the bottom level CMakeLists file as in [here](https://github.com/Intermodalics/ros_android/blob/kinetic/example_workspace/src/hello_world_example_app/app/build.gradle#L31-L36).
+- Finally, write your [bottom level CMakeLists file](https://github.com/Intermodalics/ros_android/blob/kinetic/example_workspace/src/hello_world_example_app/app/src/main/cpp/CMakeLists.txt). Note that the way of finding and adding packages as dependencies should be just as in any regular Catkin project.
```

Comment:
Plugins have to be handled specially (`-Wl,--whole-archive ...`).

We also had to add `"-Wl,--exclude-libs,ALL"` before any other library in `target_link_libraries()` when linking shared libraries to static libraries. Otherwise binaries become huge. With the default option the shared library would contain all the symbols from all the static libraries it links to, while for executables unneeded symbols are typically omitted. I have not checked yet whether this is also the case for the examples.

Reply:
Good suggestion; I'd add these clarifications for plugins in a separate section below these guidelines.

Reply:
> I have not checked yet whether this is also the case for the examples.

I tried adding `-Wl,--exclude-libs,ALL` before linking the libraries and for the examples it doesn't make a significant difference, and adding the flags made the app crash during runtime because of missing symbols. Perhaps I'm missing something? 

On the other hand, from what I saw the binary size increases a lot with debug symbols.

Reply:
No worries.

In our case, a quite complex node with many 3rd-party dependencies compiled as static libraries, adding `-Wl,--exclude-libs,ALL` made a small, but not significant difference in terms of binary size. Maybe you are right and I have mixed debug with non-debug build artifacts during my first tests. Still, the symbol table is cleaner when using `-Wl,--exclude-libs,ALL` and only the exported symbols needed for JNI (or `android_main`) are preserved.

Let's leave it as-is.